keps/sig-storage/20200319-selinux-relabeling.md

```diff
@@ -0,0 +1,477 @@
+   Kubelet passes also implicit Secret volume with token with ":Z".
+2. Container runtime uses SELinux label "s0:c10,c0", as instructed by Kubernetes. It will recursively relabels all volumes with ":Z" to this label.
+
+#### Story 3
+
+User (or something else, e.g. an admission webhook) configures SELinux label for a pod.
+User chooses `SELinuxRelabelPolicy: "OnVolumeMount"`, because they expect a potentially large volume to be used by the pod.
+
+```yaml
+apiVersion: v1
+kind: Pod
+metadata:
+  name: testpod
+spec:
+  securityContext:
+    seLinuxOptions:
+      level: s0:c10,c0
+    seLinuxRelabelPolicy: OnVolumeMount
+  containers:
+    - image: nginx
+      name: nginx
+      volumeMounts:
+        - name: vol
+          mountPath: /mnt/test
+  volumes:
+      - name: vol
+        persistentVolumeClaim:
+          claimName: myclaim
+```
+
+In this case, kubelet tries to mount all pod's volumes with `-o context=s0:c10,c0` mount option`.
+If it succeeds, it passes the volume to the container runtime without ":Z" and the container runtime does not relabel the volume.
+See [New Kubernetes behavior](#new-kubernetes-behavior) for error cases.
+
+
+
+### Implementation Details/Notes/Constraints [optional]
+
+### Risks and Mitigations
+
+## Design Details
+
+### Test Plan
+
+* Unit tests:
+   * API validation (all permutations missing / present PodSecurityPolicy.SELinuxOptions & SELinuxRelabelPolicy & container.SecurityPolicy.SELinuxOptions)
+   * Passing mount options from kubelet to volume plugins.
+* E2e tests:
+   * Check no recursive `chcon` is done on a volume when not needed /
+   * Check recursive `chcon` is done on a volume when needed (with a matrix of SELinuxOptions / SELinuxRelabelPolicy).
+* Prepare e2e job that runs with SELinux in Enforcing mode!
+
+### Graduation Criteria
+
+* Alpha:
+ * Provided all tests defined above are passing and gated by the feature gate `SELinuxRelabelPolicy` and set to a default of `false`.
+ * Documentation exists.
+* Beta: with discussions in SIG-Storage regarding success of deployments. A metric will be added to report time taken to perform a volume ownership change. Feature gate `ConfigurableFSGroupPolicy` is `true`.
+* GA: all known issues fixed.
+
+### Upgrade / Downgrade Strategy
+
+`SELinuxRelabelPolicy` becomes "invisible" or dropped in an downgraded cluster. Container runtime will get ":Z" on volumes and they will do slow recursive chown as they do today.
+
+### Version Skew Strategy
+
+
+
+## Implementation History
+
+* 1.19: Alpha
+
+## Drawbacks [optional]
+
+* This KEP changes behavior of volumes shared by multiple pods, where each of them has a different SELinux label. See [Shared Volumes](#shared-volumes) for detail.
+* The API is slightly different that `FSGroupChangePolicy`, which may create confusion.
+
+## Alternatives [optional]
+
+### `FSGroupChangePolicy` approach
+The same approach & API as in `FSGroupChangePolicy` can be used.
+This is a viable option.
```

Comment:
What advantages does the current proposal have compared to this? The main thing I see is even the first time we don't need to recursively label.

Reply:
Added advantages / disadvantages. The main drawback of the alternative approach is speed of the first volume use.